wgsl/index.bs

```diff
@@ -5690,6 +5690,363 @@ Note: If the function [=return value|returns a value=], that value is ignored.
+
+We note "*s*: *B*" to say that *s* respects the rules regarding behaviors, and has [=behavior=] *B*.
+
+For each function:
+- its body must be a valid statement by these rules
+- if it has a return type, its [=behavior=] must be one of {Return} or {Return, Discard}
+- otherwise, its [=behavior=] must be a subset of {Default, Return, Discard}
+
+We assign a [=behavior=] to each function: it is its body's [=behavior=] (treating the body as a regular statement), with any "Return" replaced by "Default".
+As a consequence of the rules above, a function behavior is always one of {Default}, {Discard} or {Default, Discard}.
+
+Similarly, we assign a [=behavior=] to each expression, since expressions can include function calls, which can discard.
+Like functions, expression behaviors are always one of {Default}, {Discard}, or {Default, Discard}.
+
+Note: There is currently no valid program with an expression that has a [=behavior=] of {Discard}.
+The reason is that only functions without a return type can have such a [=behavior=], and there is no compound expression in which such a function can be called.
+
+<table class='data'>
+  <caption>Rules for analyzing and validating the behaviors of statements</caption>
+  <thead>
+    <tr><th>Statement<th>Preconditions<th>Resulting behavior
+  </thead>
+  <tr>
+    <td class="nowrap">*s1* ; *s2*
+    <td class="nowrap">*s1*: *B1*<br>
+        Default in *B1*<br>
+        *s2*: *B2*
+    <td class="nowrap">*B1*&#x2216;{Default} &cup; *B2*
+  <tr>
+    <td class="nowrap">var x:T;
+    <td>
+    <td>{Default}
+  <tr>
+    <td class="nowrap">let x = *e*;
+    <td class="nowrap">*e*: *B*
+    <td>*B*
+  <tr>
+    <td class="nowrap">var x = *e*;
+    <td class="nowrap">*e*: *B*
+    <td>*B*
+  <tr>
+    <td class="nowrap">x = *e*;
+    <td class="nowrap">*e*: *B*
+    <td>*B*
+  <tr>
+    <td class="nowrap">f(*e_1*, ..., *e_n*);
+    <td class="nowrap">*e_1*: *B_1*<br>
+        ...<br>
+        *e_n*: *B_n*<br>
+        *f*'s body has behavior *B*
+    <td class="nowrap">*B* &cup; (*B_1* &cup; ... &cup; *B_n*)&#x2216;{Default}
+  <tr>
+    <td>return;
+    <td>
+    <td>{Return}
+  <tr>
+    <td class="nowrap">return *e*;
+    <td class="nowrap">*e*: *B*<br>
+    <td class="nowrap">*B*&#x2216;{Default} &cup; {Return}
+  <tr>
+    <td>discard;
+    <td>
+    <td>{Discard}
+  <tr>
+    <td>break;
+    <td>
+    <td>{Break}
+  <tr>
+    <td>continue;
+    <td>
+    <td>{Continue}
+  <tr>
+    <td>fallthrough;
+    <td>
+    <td>{Fallthrough}
+  <tr>
+    <td class="nowrap">if (*e*) *s1* elseif *s2* ...  else *sn*
+    <td class="nowrap">*e: B*<br>
+        *s1*: *B1*<br>
+        ...<br>
+        *sn*: *Bn*
+    <td class="nowrap">*B*&#x2216;{Default} &cup; B1 &cup; ... &cup; Bn
+  <tr>
+    <td class="nowrap">loop {*s*}
+    <td class="nowrap">*s*: *B*
+    <td class="nowrap">let b = *B*&#x2216;{Default};<br>
+        if (Break in b) b = b&#x2216;{Break} &cup; {Default};<br>
+        b&#x2216;{Continue}
```

Comment:
I don't understand this comment, to the best of my knowledge we don't have a `then` keyword?

Reply:
I may be misreading this.  Let me look again.

Reply:
I think I understand this now.  The last column is a 3-statement procedure, using 'b' as an updatable set variable.

Can we write this in two rows, with different preconditions:
 
     loop {s}           s:B, and Break in B                        B \ { Continue} union { Next }

     loop {s}           s:B, and Break not in B                  B \ {Next,Continue}

That reads more clearly to me.

I think the next loop case can be written this way too.

(We can use a rowspan=2 in the first column to avoid the ugly duplication.)

The switch can be written this way too, but the Break that takes effect only occurs in the case clauses.  
(Hmm. I guess one can prove that no expression has Break, Continue, Fallthrough, or Return behaviours. That may be worth a note.)

I also realized that some infinite loops have empty behaviours!   :-)